mevislab.github.io/content/tutorials/basicmechanisms/macromodules.md

```diff
@@ -15,16 +15,108 @@ menu:
+
+### Files Associated with a Macro Module:
+Macro modules typically contain the following files:
+* **Definition file (*\*.def*):** The module definition file contains the definition and information about the module like name, author, package, etc. **Definition files are only available for global macro modules**.
+
+* **Script file (*\*.script*):** The script file defines inputs, outputs, fields and the user interface of the macro module. In case you want to add Python code, it includes the reference to the Python script file. The *\*.script* file allows you to define Python functions to be called on field changes and user interactions.
+
+  * **Module Initialization**: You can add the *initCommand* to the *Commands* section and the given Python function is called whenever the module is added to the workspace or reloaded.
+  * **Window creation**: You can add the *initCommand* to the *Window* section and the given Python function is called whenever the panel of the module is opened.
+  * **User interaction**: You can add commands to any user interface element like *Buttons* to call Python functions on user interactions with this element. The image below shows you the user interface and the internal interface:
+  ![user interface and the internal interface](/images/tutorials/basicmechanics/mycountourFilter.png "user interface and the internal interface")
+
+  * **Field changes**: You can also react on any changes of fields in your module and create Field Listeners. See below for details.
+* **Python file (*\*.py*):** *(Optional)*: The Python file contains the Python code that is used by the module. You can add Python functions to fields or UI elements to react on user interactions in the *\*.script* file.
+* **MeVisLab network file (*\*.mlab*):** Stores the internal network of modules that make up the macro module. This file essentially defines the macro module's internal structure and connections.
+* **Macro help file (*\*.help*):***(Optional)*:Provides help documentation for the macro module. This file is used to display information to users about the module’s functionality, usage, and any specific instructions.
+### Field Listeners:
+Field listeners are mechanisms to execute Python code automatically when the value of a field changes. This allows you to create dynamic responses to user interactions in the module's parameter panel.
+
+You create field listeners within the *Commands* sections of the *\*.script* file. You get a reference to the field object and then use a method to add a callback function that will be executed when the field's value is modified.
+
+For an example see [Example 2.5.2: Module interactions via Python scripting](/tutorials/basicmechanisms/macromodules/scriptingexample2/).
+
+## Summary
+* Macro Module: a custom block in MeVisLab used to group a network of modules or custom Python code into a single reusable unit. 
```

Comment:
Macro modules allow you to add your own functionality to MeVisLab. You can add inputs and outputs and connect existing modules to your new macro module.